### PR TITLE
In-memory Engine: high performant concurrent lock-free skiplist

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Ref #16141

What's Changed:

```commit-message
high performant concurrent lock-free skiplist
```
Due to lifetime/ownership restrictions, crossbeam skiplist cannot be used directly by in-memory engine. This PR modifiy it to make it usable.

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by Connor1996 — COMMENTED** [review]
What's the main difference compared to the crossbeam's skiplist

**Comment by SpadeA-Tang** [comment]
> What's the main difference compared to the crossbeam's skiplist

In crossbeam skiplist, the iterator has borrow of skiplist, making iterator unable to `send` which is required by engine_traits::Iterator. This is why we cannot use it directly.

crossbeam skiplsit uses epoch based reclaimation algorithm to avoid the concurrency issue between read/write and delete. So for all operations including get, insert, and remove, it needs a `guard`, where the memory free operation is done lazily by guard.
Here is the signature of get, and the struct Entry.
```
pub fn get<'a: 'g, 'g, Q>(&'a self, key: &Q, guard: &'g Guard) -> Option<Entry<'a, 'g, K, V>>
    where
        K: Borrow<Q>,
        Q: Ord + ?Sized,
    {...}
pub struct Entry<'a: 'g, 'g, K, V> {
    parent: &'a SkipList<K, V>,
    node: &'g Node<K, V>,
    guard: &'g Guard,
}
```
The Node in Entry is protected by guard and will not be destroyed as long as guard exists.
But Guard is also not `send`, and using a guard this way is pretty hard to make it integrate with in-memory engine. 
Luckly, the node iteself reference counted and there is a struct to utilize it:
```
pub struct RefEntry<'a, K, V> {
    parent: &'a SkipList<K, V>,
    node: &'a Node<K, V>,
}
```
Before return the RefEntry, it increases the ref count, so it does not need guard anymore. 

All my changes are related to converting all `Entry` to `RefEntry` and replacing skiplist with &skiplist.

File: components/region_cache_memory_engine/src/skiplist/list.rs [hunk 1]
```diff
@@ -0,0 +1,1523 @@
...
+                    // Create a guard that destroys the new node in case search panics.
+                    let sg = scopeguard::guard((), |_| {
+                        Node::finalize(node.as_raw(), self.inner.memory_controller.clone())
+                    });
+                    search = self.search_position(&n.key, guard);
+                    mem::forget(sg);
+                }
+
+                if search.found.is_some() {
+                    // panic!("Overwrite is not supported, {:?}", (*r).key);
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/skiplist/list.rs:687]: Why don't we support overwrite? The original crossbeam skiplist support it.

Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/skiplist/list.rs:687]: Overwrite can cause problem, the crossbeam implements it by first deleting than put, so during the period, the data can be deemed as delete. In addition, we have sequence number, it should not be found.

**Pull request closed without merging**